src/Microsoft.AspNetCore.Pubternal.Analyzers/PubternalityAnalyzer.cs

```diff
@@ -0,0 +1,240 @@
+                analysisContext.RegisterSyntaxNodeAction(syntaxContext => AnalyzeTypeUsage(syntaxContext), SyntaxKind.IdentifierName);
+            });
+        }
+
+        private void AnalyzeTypeUsage(SymbolAnalysisContext context)
+        {
+            var ns = (INamespaceSymbol)context.Symbol;
+            if (ns.Name == "Internal")
+            {
+                return;
+            }
+
+            foreach (var namespaceOrTypeSymbol in ns.GetMembers())
+            {
+                if (namespaceOrTypeSymbol.IsType)
+                {
+                    CheckType((ITypeSymbol)namespaceOrTypeSymbol, context);
+                }
+            }
+        }
+
+        private void CheckType(ITypeSymbol typeSymbol, SymbolAnalysisContext context)
+        {
+            if (IsPrivate(typeSymbol) || IsPrivate(typeSymbol.ContainingType))
+            {
+                return;
+            }
+
+            CheckAttributes(context, typeSymbol.GetAttributes());
+
+            if (typeSymbol.BaseType != null)
+            {
+                CheckType(context, typeSymbol.BaseType, typeSymbol.DeclaringSyntaxReferences);
+            }
+
+            foreach (var interfaceImpl in typeSymbol.AllInterfaces)
+            {
+                CheckType(context, interfaceImpl, typeSymbol.DeclaringSyntaxReferences);
+            }
+
+            foreach (var member in typeSymbol.GetMembers())
+            {
+                CheckMember(context, member);
+            }
+
+            foreach (var innerType in typeSymbol.GetTypeMembers())
+            {
+                CheckType(innerType, context);
+            }
+
+            if (typeSymbol is INamedTypeSymbol namedTypeSymbol)
+            {
+                if (namedTypeSymbol.DelegateInvokeMethod != null)
+                {
+                    CheckMethod(context, namedTypeSymbol.DelegateInvokeMethod);
+                }
+            }
+        }
+
+        private void CheckMember(SymbolAnalysisContext context, ISymbol symbol)
+        {
+            if (IsPrivate(symbol))
+            {
+                return;
+            }
+
+
+            switch (symbol)
+            {
+                case IFieldSymbol fieldSymbol:
+                {
+                    CheckAttributes(context, fieldSymbol.GetAttributes());
+                    CheckType(context, fieldSymbol.Type, fieldSymbol.DeclaringSyntaxReferences);
+                    break;
+                }
+                case IPropertySymbol propertySymbol:
+                {
+                    CheckAttributes(context, propertySymbol.GetAttributes());
+                    CheckType(context, propertySymbol.Type, propertySymbol.DeclaringSyntaxReferences);
+                    break;
+                }
+                case IMethodSymbol methodSymbol:
+                {
+                    switch (methodSymbol.MethodKind)
```

Comment:
Comment that these are compiler-generated methods for other symbols we already check (events, properties, etc.)